pymc_bart/tree.py

```diff
@@ -52,77 +106,102 @@ class Tree:
         "output",
     )
 
-    def __init__(self, tree_structure, idx_leaf_nodes, output):
+    def __init__(
+        self,
+        tree_structure: Dict[int, Node],
+        idx_leaf_nodes: Optional[npt.NDArray[np.int_]],
+        output: Optional[npt.NDArray[np.float_]],
+    ) -> None:
         self.tree_structure = tree_structure
         self.idx_leaf_nodes = idx_leaf_nodes
         self.output = output
 
     @classmethod
-    def new_tree(cls, leaf_node_value, idx_data_points, num_observations, shape):
+    def new_tree(
+        cls,
+        leaf_node_value: float,
+        idx_data_points: Optional[npt.NDArray[np.int_]],
+        num_observations: int,
+        shape: int,
+    ) -> "Tree":
         return cls(
             tree_structure={
-                0: Node.new_leaf_node(0, value=leaf_node_value, idx_data_points=idx_data_points)
+                0: Node.new_leaf_node(
+                    index=0, value=leaf_node_value, idx_data_points=idx_data_points
+                )
             },
-            idx_leaf_nodes=[0],
+            idx_leaf_nodes=np.array([0]),
             output=np.zeros((num_observations, shape)).astype(config.floatX).squeeze(),
         )
 
-    def __getitem__(self, index):
+    def __getitem__(self, index) -> Node:
         return self.get_node(index)
 
-    def __setitem__(self, index, node):
+    def __setitem__(self, index, node) -> None:
         self.set_node(index, node)
 
-    def copy(self):
-        tree = {
+    def copy(self) -> "Tree":
+        tree: Dict[int, Node] = {
             k: Node(v.index, v.value, v.idx_data_points, v.idx_split_variable)
             for k, v in self.tree_structure.items()
         }
-        return Tree(tree, self.idx_leaf_nodes.copy(), self.output.copy())
+        idx_leaf_nodes = self.idx_leaf_nodes.copy() if self.idx_leaf_nodes is not None else None
+        output = self.output.copy() if self.output is not None else None
```

Comment:
When this can be None?

Reply:
thanks, I missed that. We "trim" trees before we send them to the Trace/InferenceData. We do this to reduce the memory footprint as we no longer use those trees for sampling, so we can change the Nones in `Tree(tree, None, None)` to anything we want. An empty list and array or a list with nan, or a sentinel value like -1...

Reply:
ahhh perfect! This makes sense! This will make mypy (and me) happier.

Reply:
fixed in https://github.com/pymc-devs/pymc-bart/pull/78/commits/8d4fc232cd415b0aa9384d6c5bd506fd2c94518e